### DISCUSSION TITLE
How to do multiple files batch processing?

### DISCUSSION THREAD

**Discussion by despairTK** [Purfview/whisper-standalone-win#29] [Q&A] [answered]
Since I will modify some parameters when generating video subtitles, I need to use the command line to run it. Currently, does it support multiple videos or video files in the same folder to be processed one by one?

If batch processing is possible, can you tell me how to do it.

**Comment by Purfview** [comment, marked as answer]
Every command line program "supports" it.
To be precise, there is nothing to support, just write a command for your command-line interpreter to do a batch processing.

Simplest example for Command Prompt:
`for %i in (*.mkv) do whisper-faster.exe --language=en --model=medium %i`

This will process one by one every mkv file in a current folder.

NOTE:
It can have significant overhead as it will reload executable and model on every file, so if you have many short files you may want to process them all in one go instead of one by one.

EDIT:
From 134.5 version Whisper-Faster supports batch processing by directory, filelist or wildcard out of the box.
And there is new `--batch_recursive` argument.

Few usage examples:

`whisper-faster.exe "D:\Clips\*.mkv" --language=en --model=medium -o=source`

`whisper-faster.exe "D:\Audio" --language=en --model=medium --batch_recursive`

`whisper-faster.exe "D:\Band\Album.m3u" --language=en --model=medium --vad_filter=False`

**Reply by ghost** [reply]
does whisper-faster.exe "D:\Clips\*.mkv" and whisper-faster.exe "D:\Clips" --language=en --model=medium --batch_recursive same thing?

**Reply by Purfview** [reply]
> does whisper-faster.exe "D:\Clips*.mkv" and whisper-faster.exe "D:\Clips" --language=en --model=medium --batch_recursive same thing?

No. `D:\Clips` is path to folder.

**Comment by Purfview** [comment]
Here is Command Prompt example without overhead, process all mkv files in one go.

Save the code posted bellow as text file and rename its extension to `bat`, then execute it.
```
@echo off
cd /d "%~dp0"
set list=
for %%f in (*.mkv) do call set list=%%list%% "%%f"
call :strlen "%list%", len
exit /b

:strlen string len
SetLocal EnableDelayedExpansion
set "token=#%~1" & set "len=0"
for /L %%A in (12,-1,0) do (
    set/A "len|=1<<%%A"
    for %%B in (!len!) do if "!token:~%%B,1!"=="" set/A "len&=~1<<%%A"
)
EndLocal & set %~2=%len%

echo.
echo Batch note: Files list length is %len%.
echo.

IF %len% == 0 goto ERROR1
IF %len% GTR 8000 goto ERROR2
IF %len% LEQ 8000 goto START

:START
whisper-faster.exe --language=en --model=medium %list%
echo.
echo Batch operation complete!
echo.
pause
exit

:ERROR1
echo Error: No files found.
echo [Place .bat in the same folder where files are and check the files extension in 4th line]
echo.
pause
exit

:ERROR2
REM https://stackoverflow.com/questions/29354602
echo Error: Command is too long! [limit is 8191 characters]
echo.
pause
exit
```

**Comment by ClaireCJS** [comment]
You should try TCC command line, it has no limit to command line length other than your actual RAM.

I'm going to do this on 30,000 or so files.... But I want to just use the EXE since it handles things so well.    I could definitely chunk it up, but it's nice to have no limitation.

**Comment by ClaireCJS** [comment]
That being said -- filelist support would be better. And it would solve the problem of length limits.

For example, an .m3u file is a playlist of mp3s/songs.

That would be great to run through this. Load the model once, and create karaoke for every song in the playlist with one go.

**Reply by Purfview** [reply]
Yeah, batch processing could be implemented into exe, maybe add some argument like `--batch` then supply path to folder with existing `audio` argument, then some code to make a list of all audio files.

Not sure if it's worth to bother when batching can be done simply by CMD, TCC & ect.
If you want to implement it then I don't mind. [cli in faster-whisper is almost same as in vanilla whisper]

**Comment by Purfview** [comment]
CMD and PowerShell have command length limit of 8191 characters, that's why there are error checks in the example in my second post.

[TCC](https://jpsoft.com/all-downloads/all-downloads.html) should have no limits, so below is example without error checks to run .bat  in TCC:

```
@echo off
cd /d "%~dp0"
set list=
for %%f in (*.mkv) do call set list=%%list%% "%%f"
whisper-faster.exe --language=en --model=medium %list%
pause
```

**Reply by deadcoder0904** [reply]
@Purfview is there a way to download the `.srt` in the same folder i used in `list`?

for example, i change only the variable `list` in the following `batfile`.

### transcribe.bat

```bat
@echo off
cd /d "%~dp0"
set list="E:\Tuts\ChatGPT Mastery\"
for %%f in (*.mkv) do call set list=%%list%% "%%f"
whisper-faster.exe --language=en --model=medium %list%
pause
```

i want my `.srt` in `E:\Tuts\ChatGPT Mastery\` instead but it's currently being put in the folder i run `whisper-faster.exe` in. i mean where i run `transcribe.bat` in which is the above file.

**Reply by deadcoder0904** [reply]
oops, i totally missed not using `--output_dir` from whisper. i'll try that.

thanks for this. it worked but it did make some issues after downloading tcc pro trial version. i didn't know i had to download tcc-rt version.

**Reply by deadcoder0904** [reply]
oh, i see `batch_recursive` now:

```bash
whisper-faster.exe "E:\Tuts\ChatGPT Mastery" --language=en --model=medium --batch_recursive
```

i should probably use this from now on.

**Reply by Purfview** [reply]
Yes, these batch scripts are non needed anymore, if you don't want to do recursive then you can use `-o=source`.

**Reply by deadcoder0904** [reply]
thanks, i just put it in `.bat` file just for the sake of it.

for individual folder, use this:

### subtitle-generator.bat

```bat
@echo off
cd /d "%~dp0"
set list="E:\Tuts\AI Art\ChatGPT Mastery\5 - Congratulations Next Steps"
for %%f in (*.mp4) do call set list=%%list%% "%%f"
whisper-faster.exe %list% --language=en --model=medium --output_dir=%list%
pause
```

for recursive, use this:

### subtitle-generator-recursive.bat

```bat
@echo off
cd /d "%~dp0"
set list="E:\Tuts\AI Art\ChatGPT Mastery"
for %%f in (*.mp4) do call set list=%%list%% "%%f"
whisper-faster.exe %list% --language=en --model=medium --batch_recursive
pause
```

prefer terminal tbh but windows doesn't have snippet expander like mac or atleast i haven't found it.

**Reply by Purfview** [reply]
`--batch_recursive` doesn't work like that. Forget those batch scripts.

Here is the proper usage for your examples:

`whisper-faster.exe "E:\Tuts\AI Art\ChatGPT Mastery\5 - Congratulations Next Steps\*.mp4" -l=en -m=medium -o=source`

or:

`whisper-faster.exe "E:\Tuts\AI Art\ChatGPT Mastery\*.mp4" -l=en -m=medium --batch_recursive`

**Reply by deadcoder0904** [reply]
it worked perfectly fine since yesterday.

i used it twice.

see i did change my `for` loop from your `.mkv` to `.mp4` so it worked.

although i should use your one-liner, i'd have to remember it which i don't want to so i use `.bat` scripts.

i just copy-paste the folder path i want subtitles on & change it in `.bat` file & then double-click it.

it does the job perfectly well.

**Comment by ClaireCJS** [comment]
That's exactly what I'm doing. I'll have to figure out how to do it with a filelist, but i know it's possible using only internal commands. :)

**Comment by ClaireCJS** [comment]
For what it's worth, here's a simple python pattern for taking an argument, and processing it whether it's a file, a wildcard, or a folder(recursive)

I haven't written one to take a filelist, but if the file extension isn't audio and it appears to be a filelist (easy to check, read the lines, see if any are files), it could be treated transparently in the same fashion.

```


def process_file(file_path, subdir):
    [main program guts here]

def process_files(file_paths):
    for file_path in file_paths:
        process_file(file_path)
    #optional ostprocess_files(".")

def process_directory(directory):
    for subdir, dirs, files in os.walk(directory):
        for filename in files:
            filepath = subdir + os.sep + filename
            process_file(filepath, subdir)
        #optional: postprocess_files(subdir)


def main():
    path = sys.argv[1]

    if os.path.isdir(path):                                                     # If path is a directory
        print(f"\nProcessing files in directory: {path}\n")
        process_directory(path)
    elif os.path.isfile(path):                                                  # If path is a file
        print(f"\nProcessing file: {path}\n")
        process_file(path, subdir=".")
    elif '*' in path:                                                           # If path is a wildcard
        files = glob.glob(path)
        print(f"\nProcessing files: {files}\n")
        process_files(files)
    else:                                                                       # If path is neither a file, nor a directory, nor a wildcard
        print("\nInvalid argument. Please enter a valid file path, wildcard, or directory.\n")

```

**Reply by Purfview** [reply]
It should be adapted to cli there: https://github.com/openai/whisper/blob/b91c907694f96a3fb9da03d4bbdc83fbcd3a40a4/whisper/transcribe.py#L378

**Reply by ClaireCJS** [reply]
I agree!!!!

[and that would make it so the exe you distribute supports it the, right?]

**Reply by Purfview** [reply]
Yes, it's practically same.
Check there: https://github.com/Purfview/whisper-standalone-win/blob/e0472ddaa51ca0c5ff9775dcdc589e1ab878c28f/faster-whisper%20cli/__main__.py#L220

**Comment by Purfview** [comment]
From 134.5 version it supports batch processing by directory, filelist or wildcard out of the box.
And there is new `--batch_recursive` argument

**Comment by ClaireCJS** [comment]
Quite awesome! I was literally showing my friends visiting last night how awesome it is and how fast it runs. I did an entire album in 4 minutes!

Other-whisper, it's like 10 minutes a song with cuda, and like 3 hours a song without.

To go from hours to minutes is A-M-A-Z-I-N-G ... thank you so much!!! That's the difference in making huge runs possible

Now i just gotta figure out splitting out the vocals with DEMUCS, pre-prompting whisper by downloading the lyrics with lyricsgenius.py/exe to reduce mistkaes, and any other options that might increase vocal accuracy and damn. It'll be a complete solution I've wanted for decades! Woo! thank you!

**Comment by Ed1ks** [comment]
Does it recognize if an .srt file with same name exists and skips it?

**Reply by Purfview** [reply]
> Does it recognize if an .srt file with same name exists and skips it?

What do you mean by "it"?

**Reply by deadcoder0904** [reply]
yes, it does. there is a `--skip` option for the same. if you use `--skip`, then it will skip the same name srt file.

**Comment by gkngkngkn** [comment]
@Purfview  Sir, in batch task execution(all parameters are default.), I have encountered an issue where the presence of a silent file in the batch results in a transcription error, leading to the interruption of the entire transcription process. like below

Traceback (most recent call last):
  File "D:\whisper-fast\__main__.py", line 977, in <module>
  File "D:\whisper-fast\__main__.py", line 876, in cli
  File "faster_whisper\transcribe.py", line 274, in transcribe
  File "faster_whisper\audio.py", line 52, in decode_audio
  File "faster_whisper\audio.py", line 103, in _resample_frames
  File "faster_whisper\audio.py", line 90, in _group_frames
  File "faster_whisper\audio.py", line 80, in _ignore_invalid_frames
  File "av\container\input.pyx", line 202, in decode
  File "av\container\input.pyx", line 134, in demux
  File "av\container\streams.pyx", line 117, in av.container.streams.StreamContainer.get
IndexError: tuple index out of range
[9988] Failed to execute script '__main__' due to unhandled exception!

another issue is when the transcription encounters a corrupted file that cannot be played, it also terminates the entire task.

Is there a way to generate empty subtitle files directly when meet such silent files? or automatically skip file causing transcription issues(maybe other unknow issues like a audio file is broken) and proceed to run the next file directly. to prevent the entire process from being interrupted?

 hope this can be fix in the next version. thx

**Reply by Purfview** [reply]
Use `--check_files` switch, it will check the files for errors.

> hope this can be fix in the next version. thx

You could search your error in the issues -> [IndexError: tuple index out of range](https://github.com/Purfview/whisper-standalone-win/issues?q=tuple+index+out+of+range)

**Comment by AnthonyEmiom** [comment]
It is possible in Windows to allocate global system memory which isnt attached to the calling process. 

This would mean you could load the model and perform other time consuming one off initialisation on it. The first time the exe was run.

Then process and quit the whisper exe.

 When it was run the next time it could use the model held in memory if it matched the next run arguments.  If they arguments where different it could free the global memory and start again.

Additionally, you could add an argument to the exe to flush the global memory when finished.

This would offer the speed up of batch processing without the downsides.

**Reply by Purfview** [reply]
Can you link to such Python implementation?

**Comment by ClaireCJS** [comment]
^^ That would be a great substitute for lack of batch processing.

**Reply by Purfview** [reply]
>for lack of batch processing.

What do you mean, you even wrote part of the batching code... 😆

**Reply by ClaireCJS** [reply]
> > for lack of batch processing.
> 
> What do you mean, you even wrote part of the batching code... 😆

oh my gosh, you're right! 😂😂😂😂 That's what I get for commenting after midnight.

It's good to have a laugh in the morning. thank you!
